File: internal/instructions/block.go
package instructions

import "strings"

const (
	// BeginMarker is the start delimiter for a managed block in a target file.
	BeginMarker = "<!-- symbrain:begin -->"
	// EndMarker is the end delimiter for a managed block in a target file.
	EndMarker = "<!-- symbrain:end -->"

	// EscapeSentinel prefixes every escaped reserved token in managed content.
	// It is itself encoded before the begin and end markers, making the token
	// encoding a prefix code rather than a doubling convention.
	EscapeSentinel = "<!-- symbrain:escape -->"
	// EscapedEscapeSentinel is the on-disk spelling of a literal escape sentinel.
	EscapedEscapeSentinel = EscapeSentinel + "e"
	// EscapedBeginMarker is the on-disk spelling of a literal begin marker.
	EscapedBeginMarker = EscapeSentinel + "b"
	// EscapedEndMarker is the on-disk spelling of a literal end marker.
	EscapedEndMarker = EscapeSentinel + "d"
)

// Render replaces (or appends) the managed block in target with the given
// content. The result preserves everything outside the markers verbatim.
// When no markers exist the block is appended at the end of the file.
// Running Render twice with the same content produces byte-identical output.
//
// The two block delimiters are reserved in managed content. To keep arbitrary
// instruction text idempotent, literal delimiters in content are written with
// a prefix-code escape; source content itself is never modified.
func Render(target, content string) string {
	managed := escapeManagedContent(content)
	idx := strings.Index(target, BeginMarker)
	if idx == -1 {
		if target != "" && !strings.HasSuffix(target, "\n") {
			target += "\n"
		}
		return target + BeginMarker + "\n" + managed + EndMarker + "\n"
	}

	endIdx := strings.Index(target[idx:], EndMarker)
	if endIdx == -1 {
		prefix := target[:idx]
		return prefix + BeginMarker + "\n" + managed + EndMarker + "\n"
	}

	absEndIdx := idx + endIdx + len(EndMarker)
	prefix := target[:idx]
	suffix := target[absEndIdx:]
	if len(suffix) > 0 && !strings.HasPrefix(suffix, "\n") && !strings.HasPrefix(suffix, "\r\n") {
		suffix = "\n" + suffix
	}
	return prefix + BeginMarker + "\n" + managed + EndMarker + suffix
}

func escapeManagedContent(content string) string {
	// Encode the sentinel first. Each emitted token starts with the sentinel,
	// and its one-byte suffix identifies the token kind, so all token streams
	// remain distinct even when the source already contains escape tokens.
	content = strings.ReplaceAll(content, EscapeSentinel, EscapedEscapeSentinel)
	content = strings.ReplaceAll(content, BeginMarker, EscapedBeginMarker)
	content = strings.ReplaceAll(content, EndMarker, EscapedEndMarker)
	return content
}

File: rust/symbrain-instructions/src/block.rs
use super::{
    BEGIN_MARKER, END_MARKER, ESCAPE_SENTINEL, ESCAPED_BEGIN_MARKER, ESCAPED_END_MARKER,
    ESCAPED_ESCAPE_SENTINEL,
};

/// Replace the first managed block in `existing`, or append one if absent.
///
/// This deliberately operates on bytes rather than strings. Go's reference
/// implementation accepts arbitrary string bytes, so invalid UTF-8 must not
/// be rejected or normalized by the port. Delimiters and inserted line feeds
/// are ASCII; every other byte outside the first block is copied verbatim.
/// Literal delimiters in `content` are written using the stable escaped
/// spellings so the generated block remains unambiguous and idempotent.
#[must_use]
pub fn render(existing: &[u8], content: &[u8]) -> Vec<u8> {
    let managed = escape_content(content);
    let Some(begin) = find(existing, BEGIN_MARKER) else {
        let mut result = Vec::with_capacity(existing.len() + managed.len() + 64);
        result.extend_from_slice(existing);
        if !existing.is_empty() && !existing.ends_with(b"\n") {
            result.push(b'\n');
        }
        append_block(&mut result, &managed);
        return result;
    };

    let after_begin = &existing[begin..];
    let Some(relative_end) = find(after_begin, END_MARKER) else {
        let mut result = Vec::with_capacity(begin + managed.len() + 64);
        result.extend_from_slice(&existing[..begin]);
        append_block(&mut result, &managed);
        return result;
    };

    let end = begin + relative_end + END_MARKER.len();
    let suffix = &existing[end..];
    let mut result = Vec::with_capacity(existing.len() + managed.len() + 64);
    result.extend_from_slice(&existing[..begin]);
    result.extend_from_slice(BEGIN_MARKER);
    result.push(b'\n');
    result.extend_from_slice(&managed);
    result.extend_from_slice(END_MARKER);
    if !suffix.is_empty() && !suffix.starts_with(b"\n") && !suffix.starts_with(b"\r\n") {
        result.push(b'\n');
    }
    result.extend_from_slice(suffix);
    result
}

fn append_block(result: &mut Vec<u8>, content: &[u8]) {
    result.extend_from_slice(BEGIN_MARKER);
    result.push(b'\n');
    result.extend_from_slice(content);
    result.extend_from_slice(END_MARKER);
    result.push(b'\n');
}

fn escape_content(content: &[u8]) -> Vec<u8> {
    // Encode the sentinel first. Every emitted code word starts with the
    // sentinel and has a distinct suffix, so the encoding is injective even
    // for content that already contains an escape token.
    let content = replace_bytes(content, ESCAPE_SENTINEL, ESCAPED_ESCAPE_SENTINEL);
    let content = replace_bytes(&content, BEGIN_MARKER, ESCAPED_BEGIN_MARKER);
    replace_bytes(&content, END_MARKER, ESCAPED_END_MARKER)
}

fn replace_bytes(input: &[u8], needle: &[u8], replacement: &[u8]) -> Vec<u8> {
    let mut result = Vec::with_capacity(input.len());
    let mut cursor = 0;
    while let Some(relative) = find(&input[cursor..], needle) {
        let position = cursor + relative;
        result.extend_from_slice(&input[cursor..position]);
        result.extend_from_slice(replacement);
        cursor = position + needle.len();
    }
    result.extend_from_slice(&input[cursor..]);
    result
}

fn find(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    haystack
        .windows(needle.len())
        .position(|window| window == needle)
}
